File: services/ats-system-checker/src/utils/jwt_utils.py
"""
JWT Utility Functions for ATS System
Handles JWT token validation and decoding from API Gateway
"""

import jwt
from typing import Optional, Dict, Any
from fastapi import HTTPException, status
from src.config.settings import get_settings
from src.utils.logging_config import get_logger

logger = get_logger(__name__)


class JWTUtils:
    """Utility class for JWT operations"""
    
    def __init__(self):
        """Initialize JWT utilities with settings"""
        self.settings = get_settings()
    
    def decode_token(self, token: str) -> Optional[Dict[str, Any]]:
        """
        Decode and validate JWT token from API Gateway
        
        Args:
            token: JWT token string
            
        Returns:
            Decoded token payload or None if invalid
            
        Raises:
            HTTPException: If token is invalid or expired
        """
        try:
            # For API Gateway tokens, we don't need to verify signature
            # as they come from a trusted source
            decoded = jwt.decode(
                token,
                options={"verify_signature": False},
                algorithms=["HS256"]
            )
            
            logger.info(f"Successfully decoded JWT token for user: {decoded.get('userId', 'unknown')}")
            return decoded
            
        except jwt.ExpiredSignatureError:
            logger.warning("JWT token has expired")
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Token has expired"
            )
        except jwt.InvalidTokenError as e:
            logger.warning(f"Invalid JWT token: {e}")
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid token"
            )
        except Exception as e:
            logger.error(f"Error decoding JWT token: {e}")
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Token validation failed"
            )
    
    def extract_user_info(self, token: str) -> Dict[str, Any]:
        """
        Extract user information from JWT token
        
        Args:
            token: JWT token string
            
        Returns:
            Dictionary containing user information
        """
        decoded = self.decode_token(token)
        
        if not decoded:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Could not extract user information from token"
            )
        
        # Extract user information from API Gateway token format
        user_info = {
            "user_id": decoded.get("userId"),
            "aud": decoded.get("aud", []),
            "iat": decoded.get("iat"),
            "exp": decoded.get("exp")
        }
        
        # Validate required fields
        if not user_info["user_id"]:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Token missing required user information"
            )
        
        return user_info


# Global instance
jwt_utils = JWTUtils()

File: services/ats-system-checker/src/utils/jwt_utils.ts
import jwt from "jsonwebtoken";
import { Env } from "../config/env.config";

/** Decoded JWT payload from API Gateway / auth provider */
export interface JwtPayload {
  userId?: string;
  user_id?: string;
  aud?: string[];
  iat?: number;
  exp?: number;
  [key: string]: unknown;
}

export class JWTUtils {
  private secret: string;

  constructor() {
    this.secret = Env.JWT_SECRET || Env.SECRET_KEY || "default-secret";
  }

  decode_token(token: string): JwtPayload | null {
    try {
      const decoded = jwt.decode(token, { complete: false });
      if (!decoded || typeof decoded === "string") {
        throw new Error("Invalid token");
      }
      const payload = decoded as JwtPayload;
      const userId = payload.userId ?? payload.user_id ?? "unknown";
      console.log(`Successfully decoded JWT token for user: ${userId}`);
      return payload;
    } catch (error: unknown) {
      if (error && typeof error === "object" && "name" in error && (error as { name: string }).name === "TokenExpiredError") {
        console.warn("JWT token has expired");
        throw new Error("Token has expired");
      }
      const message = error instanceof Error ? error.message : "Invalid token";
      console.warn(`Invalid JWT token: ${message}`);
      throw new Error("Invalid token");
    }
  }

  extract_user_info(token: string): {
    user_id: string;
    aud?: string[];
    iat?: number;
    exp?: number;
  } {
    const decoded = this.decode_token(token);
    if (!decoded) {
      throw new Error("Could not extract user information from token");
    }
    const user_id = decoded.userId ?? decoded.user_id;
    if (!user_id) {
      throw new Error("Token missing required user information");
    }
    return {
      user_id: String(user_id),
      aud: decoded.aud,
      iat: decoded.iat,
      exp: decoded.exp
    };
  }
}

// Global instance
export const jwt_utils = new JWTUtils();

